### ISSUE TITLE
Autotype for Linux?

### ISSUE THREAD

**Issue by Friday13th87** [anonymous1184/bitwarden-autotype#51]
Hello!
Is there any way to use that in an linux environment? 
Or do anyone know a port / solution for linux autotype with bitwarden?

thanks in advance

**Comment by anonymous1184** [comment]
I don't know of any (I don't use any Linux desktop environment).

The big selling feature of Linux, having a gazillion distributions and each of them having the ability to (possibly) use any desktop environment make this really hard. Perhaps if you try in LinuxQuestions, any StackExchange site, Bitwarden forums or a search engine in general.

AutoHotkey is basically a wrapper around the WinAPI, so it was quite simple to pull the data from `StdOut` and having it ready for any window, given that all report the same characteristics to the underlying OS. If running under Wine, it will only locate the Wine-emulated apps.

The last bit I just tested with the AHK Discord bot (running under Wine).

**Comment by RokeJulianLockhart** [comment]
@anonymous1184,

> The big selling feature of Linux, having a gazillion distributions and each of them having the ability to (possibly) use any desktop environment make this really hard. Perhaps if you try in LinuxQuestions, any StackExchange site, Bitwarden forums or a search engine in general.

Any of the big few are literally identical in every way (process initializer is https://github.com/systemd/systemd, DE is either KDE Plasma or GNOME) except packaged software and app installation package type, and even this is basically just `.rpm` versus `.deb`.

> AutoHotkey is basically a wrapper around the WinAPI, so it was quite simple to pull the data from StdOut and having it ready for any window, given that all report the same characteristics to the underlying OS. If running under Wine, it will only locate the Wine-emulated apps.

If there's a https://github.com/dotnet/runtime version of the APIs you use, you'd gain total cross-platform support immediately, and the ability for others to use your code in https://github.com/PowerShell/PowerShell scripts.

> The last bit I just tested with the AHK Discord bot (running under Wine).

That should only work with apps installed via Wine though, unfortunately. Wine is slightly more restricted in some ways than https://github.com/microsoft/WSL is, even though it uses the host FS more transparently.

**Comment by anonymous1184** [comment]
@rokejulianlockhart yes, most desktop environments can be generalized, still you need to develop and test for more than one (KDE, GNOME, Unity, Xfce, Cinnamon, LXDE, Xmonad, MATE, etc...), that alongside me not using any desktop environment for Linux makes it unappealing.

Also, yes; you can use C#\* to make this happen and broaden the compatibility with PowerShell. But right now I'm sitting on over 6 months of updates with 100+ changes and upgrades, and I haven't had the time to properly document anything or make the appropriate commits to the code.

*<sup>There is an old C# port of AutoHotkey and an ongoing effort to rewrite it, you can check the AutoHotkey Discord server for details.</sup>*

The bright side? People in my workplace are helping with the testing on a rolling-updates private repository that has helped me to squash bugs faster. I really hope to make all that public, so I can hand down the project. Hopefully the next owner will be more interested to include Linux support.
